Setup CI

LukeWeidenwalker opened the pull request:
Sets up a CI pipeline to resolve #99.

More things to do before this can be merged:
- Clean up the build + dependencies
  - Note: I removed the xarray-extras dependency, because I couldn't get it to build with everything else + the functionality we use it for now seems to be in xarray
  - [x] needs documentation
  - [x] add gdal python dependency
  - [x] add sensible build variants
- Adds a devcontainer to develop in
  - [x] Needs documentation
  - [x] git isn't properly installed, so the vcs extension doesn't work
- Adds a GH actions workflow that runs mypy and pytest for the supported Python versions.

LukeWeidenwalker commented:
@SerRichard @ValentinaHutter This PR is ready for a first review! You can see that tests & typechecking are run correctly on multiple versions of Python by inspecting the Github Action outputs! Note that while both checks currently fail, fixing them is out of scope for this PR and should be done separately.

Please especially let me know whether the instructions to setup the devcontainer are sufficiently clear - if not I'm happy to give a tour during a short call and update the docs accordingly! ;)

SerRichard commented on .devcontainer/postCreateCommand.sh:
```diff
@@ -0,0 +1,3 @@
+mkdir .venv
```
`#!/bin/bash`
missing at start of file

SerRichard commented on noxfile.py:
```diff
@@ -0,0 +1,13 @@
+@session(python=python_versions)
+def mypy(session):
+    session.install("mypy")
+    session.run("mypy")
```
Newline

SerRichard commented on .devcontainer/postCreateCommand.sh:
```diff
@@ -0,0 +1,3 @@
+mkdir .venv
+mv /opt/pysetup/.venv/* .venv
+poetry install
```
Wasn't aware new lines were part of the bash convention, but perhaps github is right

SerRichard commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,41 @@
+
+	// Uncomment to connect as a non-root user if you've added one. See https://aka.ms/vscode-remote/containers/non-root.
+	// "remoteUser": "vscode"
+}
```
Newline

SerRichard commented on README.md:
```diff
@@ -0,0 +1,44 @@
+- poetry for dependency management
+- nox and older versions of python to test against 
+
+This devcontainer is intended to be used in conjunction with the [`VSCode Remote - Containers`](https://code.visualstudio.com/docs/remote/containers) extension, which lets you use it as a full-featured development environment. Note that this isn't a strict requirement, but a strong recommendation that will make life easier. Also note, that this package is distributed in the standard formats throu ...
```
I have not used the VSCode Remote containers, does what you have written here require no further knowledge than what is already included in the VSCode how to guides?

SerRichard commented on noxfile.py:
```diff
@@ -0,0 +1,13 @@
+    session.run("pytest")
+
+@session(python=python_versions)
+def mypy(session):
```
As mypy is only checking static typing, it may be worth adding Flake8 for stylistic enforcement (definitely useful if the aim of doing this CI work is to make dev easier for people from different places). Though this will be horrible the first time it's ran.

LukeWeidenwalker commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,85 @@
+    apt-get install -qy --no-install-recommends \
+        git \
+        openssh-client \
+        python3.5 \
```
I've not included more versions for now to save CI minutes, but I think when this is done, we want to test 3.6 - 3.10!

LukeWeidenwalker commented:
Alright, I've now updated the devcontainer such that the UID/GID that's used to create stuff in the container can be configured and is the same as the host system - got quite a few permission denied problems on the mounted sourcecode files otherwise. 

Have you had a chance yet to try out the devcontainer bit? Would love to find out whether this works for you!

SerRichard commented:
Do you know what is causing the latest failure in the cotnainer build? Some permissions issue it looks like?

SerRichard commented on README.md:
```diff
@@ -0,0 +1,56 @@
+- (only on pushes to master) If all checks pass, publish the devcontainer to the container registry.
+
+Because the CI uses the devcontainer as the base environment to run nox sessions, they can also be run locally from the devcontainers shell, e.g. 
+`nox --session tests --python=3.8` to run the tests session on a specific version of python.
```
This will not work currently if someone uses the docker build locally, just a note that when running locally the dir needs to be mounted for the nox tests to run. As I didn't notice this when I ran locally

SerRichard reviewed (commented):
Only one comment now, I think we resolved the other comments!

LukeWeidenwalker commented:
> Do you know what is causing the latest failure in the cotnainer build? Some permissions issue it looks like?

Huh, so far I thought that was because the workflow is really running on my fork, but actually I think this is to do with the permissions on the package registry of the Open-EO org. 

@m-mohr sorry to bother, but do you know whether Github Actions are allowed to push images to the Open-EO package registry? If not, would be great if this could be enabled to let us publish the devcontainer & use the docker cache in CI - thanks!

LukeWeidenwalker commented:
@soxofaan @clausmichele 
Hey both - I've recently been working on setting up a CI pipeline for this repo. Now that the basic flow works and @SerRichard has helped me iron out the obvious flaws, I think these changes are ready to be looked upon/tried out by both of you. 

Some notes:
- I've been using docker version 20.10.15 to build this
- I will extend the range of Python versions to be tested across all of 3.6-3.10, just wanted to save some compute while I was still experimenting
- I'll add proper linting in a separate PR
- Some tests and type checking turned out to be currently failing - I'm currently skipping those, just to get to an initial green build, but we'll have to fix these asap after this is merged.
- I've recently taken a liking to poetry and have been using it here - I like how all the settings are in a single file, that there's well-defined dev-dependencies, that it resolves fast and most importantly that it will make it much easier to publish this package to PyPI later! If you can't be persuaded into using it, we can look at options to include a frozen `requirements.txt`. I'd like the source of truth for dependencies to be managed by poetry though, for the reasons outlined above.
- I realize that using docker to setup GDAL adds complexity. I also realize that you might not have used the Remote Containers extension, or even be using VSCode for development at all. However, I've come at this from various angles already and as long as this repo depends on GDAL I truly think that just running everything in docker is the best option to ensure a consistent build across contributors and CI runs. It's also quite nice to use once setup. If you run into any troubles, please feel free to ping me, I'm more than happy to jump on as many calls necessary to help through any setup troubles (and thereby also improve the instructions in the readme)! 
- Are there any further build variants you'd like to see included here? We were thinking about adding one for the dask worker specifically, any other suggestions?

clausmichele commented:
Hi @LukeWeidenwalker! Thanks for the effort, really like to see this going on. I'm not an expert in setting up CI pipelines, so I can't help you much here.

About your last question: it would be nice to have a version which does not depend on the EODC set-up. For example, the save_result process currently has hard coded that the data in ODC should be stored in EQUI7:
https://github.com/Open-EO/openeo-processes-python/blob/7407d8d28433978aa47749707e2122ba0f709c11/src/openeo_processes/utils.py#L349
https://github.com/Open-EO/openeo-processes-python/blob/7407d8d28433978aa47749707e2122ba0f709c11/src/openeo_processes/utils.py#L324

clausmichele commented:
Ok I had a look at the README file. So, if you merge this we would be forced to use Visual Studio for updating/developing the repo? Or just for running the tests? Wouldn't be possible to include them as a GitHub action?

soxofaan commented:
Looks interesting! A couple of notes:

- as mentioned, I see a lot of VSCode references, e.g. also in user names in the docker container, which looks weird to me. 
- do I understand correctly that the image build is part of the test github actions workflow? It looks like a non-trivial build because you install 5 python versions :smile: . Isn't that going to be a serious time drain? 
- you also drop the `extras_require` feature (optional dependencies), right? So the package unconditionally depends on dask, odc-algo, ...  That's not ideal if we want to use this library from the VITO backend, just for numpy implementation and don't want to drag in dependencies we don't use

LukeWeidenwalker commented:
1) Ah, fair point - the important parts are the UIDs anyways, so I'll change the USERNAME to the more generic `ubuntu`!
2) The image build is indeed part of the test workflow, but I've added layer caching, so layers will only have to be rebuilt when they've actually changed!
3) Right - this is a good point too! So the desired behaviour would be to be able to do something like `pip install openeo-processes[dask]`, `pip install openeo-processes[odc]` or `pip install openeo-processes[bare]`, right? I haven't tried this yet, but something similar should be feasible with poetry dependency groups - which will only become available with poetry 1.2 (see [this issue](https://github.com/python-poetry/poetry/pull/4260)). I've created #182 and added optional dependencies as a sub-task there. Maybe by the time I get a chance to pick this up, 1.2 will be out already ;)

soxofaan commented:
> So the desired behaviour would be to be able to do something like pip install openeo-processes[dask], pip install openeo-processes[odc] or pip install openeo-processes[bare], right? I haven't tried this yet, but something similar should be feasible with poetry dependency groups ```

Indeed something like that. With the difference that a bare install would just be `pip install openeo-processes`, without more specification.

I don't know poetry very well, but it seem that it already supports "extras" (which is the feature I referred to): https://python-poetry.org/docs/pyproject/#extras

soxofaan commented:
> Note however, that unless everything is installed, the test suite won't run because src/openeo_processes/__init__.py attempts to import the entire library using from openeo_processes.arrays import * statements.

If that is the case, then it doesn't make sense to have optional dependencies. Anytime you do any sub-import like  `import openeo_processes.foo.bar`  you implicitly do an import  of `src/openeo_processes/__init__.py`  first, so nothing will work unless all the dependencies of `src/openeo_processes/__init__.py` are available.  

> Idk how you've been using numpy-only processes previously, 

We are pinned to a very old version 0.0.4 that does work without having and optional dependency like  dask installed: https://github.com/Open-EO/openeo-python-driver/blob/master/setup.py#L44

There are several approaches to handling optional dependencies:
- local "just in time" imports (instead of global module level imports)
- falling back on None, e.g. 
    ```
    try:
        import dask
    except ImportError:
        dask = None
    ```
